README.md
# Escalate HackerOne Reports to Pivotal Tracker

This Heroku app lets you click "Escalate" on a HackerOne Report and add it as a bug to Pivotal Tracker.

## Step 1

Deploy your own copy to Heroku:

[![Deploy](https://www.herokucdn.com/deploy/button.svg)](https://heroku.com/deploy?template=https://github.com/urbandictionary/hackerone-pivotaltracker)

## Step 2

Email support@hackerone.com:

```
Hello,

I'd like to add a custom escalation integration with these URLs:

https://YOUR-HEROKU-APP-NAME.herokuapp.com/create?id=[report id]&url=[report url]&title=[title]&details=[details]
https://YOUR-HEROKU-APP-NAME.herokuapp.com/view?id=[id]

Thank you!
```

Wait for HackerOne to confirm that the integration was added.

## Step 3

On a HackerOne Report, choose Change State > Triaged, then click Escalate

## Step 4

After the Pivotal Tracker story is created, paste the Pivotal Tracker story ID into the "Reference ID" input box on the HackerOne Report. That is necessary for the "View" link.

---

# Development

## Run tests

`lein midje :autotest`

## Set up development

cat >profiles.clj

```
{:dev {:env {:dev                       "true"
             :pivotaltracker-api-key    "aaaaaaa"
             :pivotaltracker-project-id "11111111"}}}
```

## Run locally

`lein run`, then open `http://localhost:8080/create?id=183837&url=https%3A%2F%2Fhackerone.com%2Freports%2F1111&title=Race+condition&details=foobar`

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-11-27
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-11-27
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/hackerone-pivotaltracker/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/hackerone-pivotaltracker/compare/0.1.0...0.1.1


test/hackerone_pivotaltracker/core_test.clj
(ns hackerone-pivotaltracker.core-test
  (:require [midje.sweet :refer :all]
            [environ.core :refer [env]]
            [org.httpkit.fake :refer [with-fake-http]]
            [ring.mock.request :as mock]
            [ring.middleware.params :refer [wrap-params]]
            [ring.middleware.keyword-params :refer [wrap-keyword-params]]
            [hackerone-pivotaltracker.core :refer :all]))

(facts "tracker-api-key"
       (fact "from env var"
             (tracker-api-key) => "foobar"
             (provided (env :pivotaltracker-api-key) => "foobar"))

       (fact "missing"
             (tracker-api-key) => (throws "Tracker API key is not set")
             (provided (env :pivotaltracker-api-key) => nil)))

(facts "tracker-project-id"
       (fact "from env var"
             (tracker-project-id) => 1234
             (provided (env :pivotaltracker-project-id) => "1234"))

       (fact "invalid"
             (tracker-project-id) => (throws "For input string: \"asdf\"")
             (provided (env :pivotaltracker-project-id) => "asdf")))

(facts "create-tracker-story"
       (let [request {:url     "https://www.pivotaltracker.com/services/v5/projects/1234/stories"
                      :headers {"Content-Type" "application/json" "X-TrackerToken" "foobar"}
                      :method  :post}]
         (against-background [(env :pivotaltracker-api-key) => "foobar"
                              (env :pivotaltracker-project-id) => "1234"]

                             (fact "sends request and parses the response"
                                   (with-fake-http [request {:status 200 :body "{\"a\":\"b\"}"}]
                                                   (create-tracker-story {:foo "bar"}) => {"a" "b"}))

                             (fact "raises when response is an error"
                                   (with-fake-http [request {:status 400 :body "{\"a\":\"b\"}"}]
                                                   (create-tracker-story {:foo "bar"}) => (throws clojure.lang.ExceptionInfo "Request failed"))))))

(fact "params->tracker"
      (params->tracker {:title "a" :details "b"}) =>
      {:description "b" :labels [{:name "hackerone"}] :name "a" :story_type "bug"})

(fact "app"
      (against-background [(env :pivotaltracker-api-key) => "foobar"
                           (env :pivotaltracker-project-id) => "1234"]
                          (fact "/create redirects to the new story URL"
                                (with-fake-http [{:url     "https://www.pivotaltracker.com/services/v5/projects/1234/stories"
                                                  :headers {"Content-Type" "application/json" "X-TrackerToken" "foobar"}
                                                  :method  :post} {:status 200 :body "{\"url\":\"http://example.com/\"}"}]
                                                (app (mock/request :get "/create?title=a&details=b")) =>
                                                {:body "" :headers {"Location" "http://example.com/"} :status 302}))

                          (fact "/view redirects to the URL of an existing story"
                                (app (mock/request :get "/view" {:id 5})) =>
                                {:body "" :headers {"Location" "https://www.pivotaltracker.com/n/projects/1234/stories/5"} :status 302})))

project.clj
(defproject hackerone-pivotaltracker "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.8.0"]
                 [http-kit "2.2.0"]
                 [http-kit.fake "0.2.1"]
                 [ring/ring-mock "0.3.0"]
                 [environ "1.1.0"]
                 [compojure "1.5.1"]
                 [ring/ring-devel "1.5.0"]
                 [prone "1.1.4"]
                 [org.clojure/data.json "0.2.6"]
                 [org.clojure/tools.logging "0.3.1"]
                 [midje "1.8.3"]]
  :min-lein-version "2.6.1"
  :plugins [[lein-environ "1.1.0"]
            [lein-midje "3.2.1"]]
  :main ^:skip-aot hackerone-pivotaltracker.core
  :target-path "target/%s"
  :profiles {:uberjar {:aot :all}})


src/hackerone_pivotaltracker/core.clj
(ns hackerone-pivotaltracker.core
  (:gen-class)
  (:require [compojure.core :refer [GET defroutes]]
            [compojure.route :as route]
            [org.httpkit.server :as server]
            [org.httpkit.client :as client]
            [clojure.data.json :as json]
            [ring.util.response :refer [redirect]]
            [ring.util.response :refer [status]]
            [prone.middleware :refer [wrap-exceptions]]
            [ring.middleware.params :refer [wrap-params]]
            [ring.middleware.keyword-params :refer [wrap-keyword-params]]
            [ring.middleware.reload :refer [wrap-reload]]
            [environ.core :refer [env]]
            [clojure.tools.logging :as log]
            [crypto.equality :as crypto]))

(defn tracker-api-key []
  (or (env :pivotaltracker-api-key)
      (throw (RuntimeException. "Tracker API key is not set"))))

(defn app-secret-key []
  (or (env :app-secret-key)
      (throw (RuntimeException. "Application secret key is not set"))))

(defn tracker-project-id []
  (Integer/parseInt (env :pivotaltracker-project-id)))

(def server-opts
  {:port (Integer/parseInt (env :port "8080")) :max-line (* 1024 64)})

(defn tracker-request-headers []
  {"Content-Type" "application/json" "X-TrackerToken" (tracker-api-key)})

(defn create-url []
  (format "https://www.pivotaltracker.com/services/v5/projects/%d/stories" (tracker-project-id)))

(defn story-url [id]
  (format "https://www.pivotaltracker.com/n/projects/%d/stories/%d" (tracker-project-id) (Integer/parseInt id)))

(defn parse-response [{:keys [status body] :as response}]
  (if (= 200 status)
    (json/read-str body)
    (throw (ex-info "Request failed" response))))

(defn create-tracker-story [attrs]
  (->> {:headers (tracker-request-headers) :body (json/write-str attrs)}
       (client/post (create-url))
       deref
       parse-response))

(defn params->tracker [{:keys [title details]}]
  {:name        title
   :description details
   :labels      [{:name "hackerone"}]
   :story_type  "bug"})

(defroutes routes
           (GET "/create" request
                (-> request
                    :params
                    params->tracker
                    create-tracker-story
                    (get "url")
                    redirect))
           (GET "/view" request
                (-> request
                    :params
                    :id
                    story-url
                    redirect))
           (route/not-found "Not Found"))

(defn verify-app-secret [app]
 (fn [req]
   (if (crypto/eq? (get-in req [:params :secret]) (app-secret-key)) (app req) (status req 403))))

(def app
  (cond-> #'routes
          ;; note: verify-app-secret has to be loaded after the
          ;; wrap-keyword-params and wrap-params middleware due
          ;; to the `params` dependency in verify-app-secret
          ;; middleware, changing the order here will cause the
          ;; get-in to return nil without a hard fail.
          true verify-app-secret
          true wrap-keyword-params
          true wrap-params
          (env :dev) wrap-reload
          (env :dev) wrap-exceptions))

(defn -main
  [& args]
  (log/info server-opts)
  (server/run-server app server-opts))
